### PR TITLE
cmd/bosun: promote scalars to numberSets

### PR DESCRIPTION
**Pull request by mjibson**

### CODE REVIEW

**Comment by mjibson** [comment]
Don't merge yet: a few more functions will get promotion. But ready to review.

**Comment by mjibson** [comment]
Other functions (des, dropge, drople, limit, nv) can also benefit from scalar promotion. But since they are not reduction functions we'll need another way to do the tag matching. They may come in a later PR. At least we get forecastlr and percentile now.

**Comment by kylebrandt** [comment]
drop__ could add a fair amount of power I think, something like:

```
$series = q("...")
$limit = last(q("..."))
$filteredSeries = dropge($series, percentille($series, .95))
$daysLeft = forecastlr($filteredSeries, $limit) / d("1d")  
``` 

Unless there is another way to do this...

**Comment by mjibson** [comment]
Yes, they will happen. There's just no obvious common way to do them all now, since they don't have the same function signature.

**Comment by mjibson** [comment]
Ok, now does the drop functions.

**Comment by kylebrandt** [comment]
docs need to be updated to reflect the drop funcs

**Comment by mjibson** [comment]
Updated with dropg, dropl, and docs.

**Comment by kylebrandt** [comment]
Will fix https://github.com/bosun-monitor/bosun/issues/1064

File: cmd/bosun/expr/expr_test.go [hunk 1]
```diff
@@ -224,3 +224,31 @@ func TestQueryExpr(t *testing.T) {
 		}
 	}
 }
+
+func TestScalarPromotion(t *testing.T) {
+	tests := map[string]map[string]Value{
```
Comment (captncraig) [inline cmd/bosun/expr/expr_test.go]: this is a really confusing data structure to only hold one test. Either add more or inline the fields (or make an anonymous struct for test cases).

File: cmd/bosun/expr/expr_test.go [hunk 2]
```diff
@@ -224,3 +224,31 @@ func TestQueryExpr(t *testing.T) {
...
+		`abs(-1)`: {"": Number(1)},
+	}
+
+	for exprText, expected := range tests {
+		e, err := New(exprText)
+		if err != nil {
+			t.Fatal(err)
+		}
+		results, _, err := e.Execute(nil, nil, nil, nil, nil, queryTime, 0, false, nil, nil, nil)
```
Comment (captncraig) [inline cmd/bosun/expr/expr_test.go]: can this exercise walkfunc directly?

File: cmd/bosun/expr/expr.go [hunk 3]
```diff
@@ -594,6 +594,15 @@ func (e *State) walkFunc(node *parse.FuncNode, T miniprofiler.Timer) *Results {
 			default:
 				panic(fmt.Errorf("expr: unknown func arg type"))
 			}
+			if f, ok := v.(float64); ok && node.F.Args[i] == parse.TypeNumberSet {
+				v = &Results{
```
Comment (captncraig) [inline cmd/bosun/expr/expr.go]: should this use the fromScalar helper?

File: cmd/bosun/expr/expr_test.go (hunk 2 above)
Comment (mjibson) [inline cmd/bosun/expr/expr_test.go]: I'm not convinced that's a great test. The scalar promotion stuff does things with groups, and I want to make sure it works at the top level.

**Comment by mjibson** [comment]
Updated with better tests - now combined with the existing expr tests, it just knows about types now.

**Pull request merged**